src/legacy/server/capabilities/resolve_capabilities.ts

```diff
@@ -0,0 +1,38 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { Request } from 'hapi';
+
+import { Capabilities } from '../../../core/public';
+import { mergeCapabilities } from './merge_capabilities';
+import { CapabilitiesModifier } from './capabilities_mixin';
+
+export const resolveCapabilities = async (
+  request: Request,
+  modifiers: CapabilitiesModifier[],
+  ...capabilities: Array<Partial<Capabilities>>
+) => {
+  let resolvedCaps = mergeCapabilities(...capabilities);
```

Comment:
If you want to write this without the mutable variable, it could be done with:

```js
=> modifiers.reduce(
  async (resolved, provider) => await provider(request, resolvedCaps),
  mergeCapabilities(...capabilities)
);
```

Or without the async function:

```js
=> modifiers.reduce(
  (promise, provider) => promise.then(resolvedCaps => provider(request, resolvedCaps)),
  Promise.resolve(mergeCapabilities(...capabilities))
);
```